Handle empty result sets in CTEs

camuthig opened the pull request:
If Django finds a CTE during SQL generation that will result in empty results, it will throw an EmptyResultSet immediately. Because the CTEs are compiled to SQL prior to the base query in the CTEQueryCompiler, though, the base compiler is missing core information Django needs to gracefully handle this situation, like `col_count` and `klass_info`.

To resolve this, in the case of EmptyResultSet being thrown by CTE compilation, the base `as_sql` will be called as well before reraising.

The creation of the base queryset can not be move before the explain behaviors in `generate_sql` or the generated `EXPLAIN` will throw an error, so the try/except pattern is used instead.

Resolves #84 

I think this will also work for #64 as well, but I couldn't figure out a way to throw that particular error.

millerdev commented:
The wrong result is returned when using `LEFT OUTER JOIN` on the CTE. Here's a test to demonstrate:
```py
    def test_left_outer_join_on_empty_result_set_cte(self):
        totals = With(
            Order.objects
            .filter(id__in=[])
            .values("region_id")
            .annotate(total=Sum("amount")),
            name="totals",
        )
        orders = (
            totals.join(Order, region=totals.col.region_id, _join_type=LOUTER)
            .with_cte(totals)
            .annotate(region_total=totals.col.total)
            .order_by("amount")
        )

        self.assertEqual(len(orders), 22)
```

camuthig commented:
@millerdev I believe the latest commit should handle that case. It requires more introspection to handle the join, since the join aspect isn't what is throwing the `EmptyResultSet` - that is coming from the CTE itself. I leveraged the aliases map to understand how the CTE is being used in the rest of the query, and then the existing `elide_empty` parameter on the `get_compiler` function to not throw those errors if those CTEs were left joins outer joins.

millerdev commented on django_cte/query.py:
```diff
@@ -75,9 +78,22 @@ def generate_sql(cls, connection, query, as_sql):
                 _ignore_with_col_aliases(cte.query)
-            compiler = cte.query.get_compiler(connection=connection)
+
+            elide_empty = True
+            alias = query.alias_map.get(cte.name)
+            if isinstance(alias, QJoin) and alias.join_type == LOUTER:
+                elide_empty = False
```
Why is this check needed? Would it work to always pass `elide_empty=False` in this context? Is there a disadvantage to doing that?

camuthig replied:
I'm kind of going with the flow of what Django does as default behavior in the core compiler. The default behavior in the compiler is to raise an assertion error and not run the query in the case of a knowable empty result set. The alternate flow, with `elide_empty` set to false, is used when queries need to be run regardless of the fact that they will not return results.

millerdev commented:
Tests are failing. Unfortunately `elide_empty` is not available on all supported versions of Django. There is also a lint error.

camuthig commented:
It looks like `elide_empty` was added in [4.0](https://github.com/django/django/commit/f3112fde981052801e0c2121027424496c03efdf). 

I have fixed the lint errors and cleaned up the code a bit to enable specifically handling the situation in the `except` block for the older versions of Django. This is basically the same way that `elide_empty` works under the hood, it just isn't as clean of an implementation in the library-space.

millerdev commented on django_cte/query.py:
```diff
@@ -75,9 +79,42 @@ def generate_sql(cls, connection, query, as_sql):
+                    # only required for left outer joins, as standard inner
+                    # joins should be optimized and raise the EmptyResultSet
+                    query = cte.query.copy()
+                    query.where = WhereNode(
+                        [ExtraWhere(["1 = 0"], [])], AND
+                    )
```
The `AND` parameter appears to be optional and the default. I didn't check older versions of Django, so maybe this changed within the support window too?
```suggestion
                    query.where = WhereNode([ExtraWhere(["1 = 0"], [])])
```

millerdev commented on django_cte/query.py:
```diff
@@ -2,15 +2,19 @@
     SQLUpdateCompiler,
 )
+from django.db.models.sql.constants import LOUTER
+from django.db.models.sql.where import ExtraWhere, AND, WhereNode
```
`AND` can be removed if `WhereNode` of older Django's does not require it.

millerdev commented on django_cte/query.py:
```diff
@@ -75,9 +79,42 @@ def generate_sql(cls, connection, query, as_sql):
-            compiler = cte.query.get_compiler(connection=connection)
+
+            alias = query.alias_map.get(cte.name)
+            is_left_outer = (
+                isinstance(alias, QJoin) and alias.join_type == LOUTER
+            )
+
+            if django.VERSION >= (4, 0):
+                compiler = cte.query.get_compiler(
+                    connection=connection, elide_empty=not is_left_outer
```
What do you think of making `is_left_outer` more generic/applicable to this context?
```suggestion
            should_elide_empty = (
                not isinstance(alias, QJoin) or alias.join_type != LOUTER
            )

            if django.VERSION >= (4, 0):
                compiler = cte.query.get_compiler(
                    connection=connection, elide_empty=should_elide_empty
```

millerdev commented on django_cte/query.py:
```diff
@@ -75,9 +79,42 @@ def generate_sql(cls, connection, query, as_sql):
+            try:
+                cte_sql, cte_params = compiler.as_sql()
+            except EmptyResultSet:
+                if django.VERSION < (4, 0) and is_left_outer:
```
```suggestion
                if django.VERSION < (4, 0) and not should_elide_empty:
```

millerdev reviewed (commented):
Looking good. I made a few suggestions.

camuthig commented:
Thanks for the feedback. It looks like the connector has always been optional, so I have removed it as requested and pulled in the other feedback.

millerdev commented on django_cte/raw.py:
```diff
@@ -32,7 +32,7 @@ def quote_name_unless_alias(self, name):
         class query(object):
             @staticmethod
-            def get_compiler(connection):
+            def get_compiler(connection, *args, **kwargs):
```
I assume this is to allow `elide_empty` to be passed without causing a `TypeError`.

What do you think of changing this to allow/ignore specific parameters rather than having it ignore all? The danger being that it could potentially ignore some future parameter that should not be ignored. In other words, I'd like to keep this interface surface area as small as possible, and have it raise errors for unexpected scenarios.
```suggestion
            def get_compiler(connection, *, elide_empty=None):
```

millerdev replied:
This is a good question. Until now I am not remembering any requests for `using`, so not sure if we need to add it. `elide_empty` is clearly needed because tests fail without it. I think the suggestion is fine for now. We can always add `using` later if needed.

millerdev reviewed (commented):
Sorry for dragging this out, but I noticed one more thing that I'd like to see changed. Otherwise this is looking great. Thanks for contributing to Django CTE!

millerdev commented:
Thank you for the contribution @camuthig!

camuthig commented:
No problem. Thanks for the awesome project @millerdev . My team is starting to lean on it more to improve our more out of control queries. Do you know when you might cut a release that includes this change? We are pinned to my branch until we can get back onto official releases, but we have a few behaviors that fall over without this fix.

millerdev commented:
Thanks for the prompt. New release: https://pypi.org/project/django-cte/1.3.3/

SupImDos commented:
@millerdev Not a huge deal, but would it be possible to do a tag / release on GitHub for `v1.3.3` as well?

millerdev commented:
Yes, it had been tagged, but I forgot to push it. https://github.com/dimagi/django-cte/releases/tag/v1.3.3
